ENH: show dtype in array repr when endianness is non-native

molsonkiko opened the pull request:
See #19059.
Goal is to get
```py
>>> import numpy as np
>>> np.array([1], dtype='>u2') # nonstandard endianness should be reflected in repr
array([1], dtype='>u2')
>>> np.array([1], dtype='<u2')
array([1], dtype=uint16)
```
<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

molsonkiko committed:
show endianness when nonstandard: address #19059
Try to fix problem where, for example,
np.array([1], dtype='>u2') and np.array([1], dtype='<u2')
both got represented as np.array([1], dtype=uint16)

seberg commented on numpy/core/arrayprint.py:
```diff
@@ -1428,6 +1428,9 @@ def dtype_is_implied(dtype):
     if dtype.names is not None:
         return False
+    
+    if dtype.byteorder != '=' and dtype.itemsize != 1:
```
Tests are failing, I suspect `dtype.isnative` should work better?

molsonkiko replied:
That works much better! I tried running `numpy/core/tests/test_arrayprint.py` with the changes in my last commit and now all the tests in that file pass.

molsonkiko committed:
see if dtype.isnative works to show endianness
@seberg pointed out that dtype.isnative could be used in place
    of my previous composite check.
Now all tests in core/tests/test_arrayprint.py pass
    (except the one xfail)

molsonkiko committed:
reformat code with shed

molsonkiko committed:
Revert "reformat code with shed"
This reverts commit eeaa2a9691251a8d504d3e2b00fc7c4699150e5b.
I wanted to fix the linter errors, but reformatting
with shed didn't help, so I'll just do it manually.

molsonkiko committed:
fix lint bugs, use pytest.mark.parameterize

molsonkiko committed:
fig bug in test_arrayprint, remove extra line

eric-wieser commented on numpy/core/arrayprint.py:
```diff
@@ -1453,10 +1457,14 @@ def dtype_short_repr(dtype):
+        # deal with cases like dtype('<u2') that are identical to an
+        # established dtype (in this case uint16)
+        # except that they have a different endianness.
+        return repr(dtype.descr[0][1])
```
Using `descr` feels like a mistake to me; `descr` is generally poorly behaved. I thin kthere should be another way to get this information?

eric-wieser replied:
`_construction_repr(..., short=True)` from https://github.com/numpy/numpy/blob/main/numpy/core/_dtype.py looks like it might do the right thing.

molsonkiko replied:
I'll look into that. `descr` was a very ad-hoc choice made because it looked like the right kind of thing.

seberg replied:
Yeah, the default in this branch looks like using `str()` or `repr()` one of them should be right (plausibly both, because complicated dtypes are dealt with earlier).

molsonkiko replied:
Given that all tests pass when using `_construction_repr(dtype)` in that conditional branch (the current status as of my most recent commit), should I leave as is, or do you think it's better to switch back?

seberg replied:
That seems still very roundabout (maybe it is actually more direct, but using more internals), and `str` is already being used for structured dtypes?

eric-wieser commented on numpy/core/tests/test_arrayprint.py:
```diff
@@ -796,6 +797,46 @@ def test_dtype_linewidth_wrapping(self):
             array(['1', '1', '1', '1', '1', '1', '1', '1', '1', '1', '1', '1'],
                   dtype='{}')""".format(styp)))
 
+    @pytest.mark.parametrize(
```
Won't this test fail on big-endian systems?

molsonkiko replied:
Yep, good point.
I think what I'll do is
1. rename `little_end` to `native` and `big_end` to `non_native`
```py
import sys
if sys.byteorder == 'big':
    # flip native and non-native for big-endian systems
    non_native = non_native.replace('>', '<')
    native = native.replace('<', '>')
```

molsonkiko committed:
change repr of non-native dtypes, fix tests
1. per eric-wieser's suggestion, using
    _construction_repr(dtype) instead of dtype.descr[0][1]
    to determine how to show non-native dtypes
2. Fix tests so they don't crash on big-endian systems.
3. Remove '<U' and '>U' tests because those are coerced
    into '<U1' and '>U1' respectively by np.array.

tylerjereddy reviewed (commented):
I tried testing on one of the compile farm Big Endian machines, but ran into some toolchain issues unrelated to the PR (ancient `gcc` version, etc.).

molsonkiko committed:
improve test coverage and variable naming

molsonkiko committed:
change _construction_repr(dtype) to str(dtype)

molsonkiko commented:
Is there more that I need to do to move this forward, or are you all just occupied with more pressing things right now?

mattip commented:
close/reopen to restart circleci, which seems to have a temporary failure.

seberg commented on numpy/core/tests/test_arrayprint.py:
```diff
@@ -796,6 +797,54 @@ def test_dtype_linewidth_wrapping(self):
+        if non_native_dtype.itemsize > 1:
+            # if the type is >1 byte, the non-native endian version
+            # must show endianness.
+            assert non_native_repr != native_repr,\
```
I much prefer avoiding `\` if possible.  In this case, I personally wouldn't care about the assert text anyway (pytest failures tend to print out the code anyway).

seberg reviewed (commented):
Just retriggered CircleCI once more.  We should make sure it is happy, since this might have modified output.

The tests could be simplified a bit, but it's OK if not.

molsonkiko committed:
simplify tests

seberg commented on numpy/core/tests/test_arrayprint.py:
```diff
@@ -796,6 +797,49 @@ def test_dtype_linewidth_wrapping(self):
+        assert ('dtype' in native_repr) ^ (native_dtype in _typelessdata),\
+                ("an array's repr should show dtype if and only if the type "
+                 'of the array is NOT one of the standard types '
+                 '(e.g., int32, bool, float64).')
```
I would prefer this as an `if native_dtype in [np.int_, np.float64, np.complex128, np.bool_]` (or `_tyeplessdata`), but that is just me nitpicking (xor is much harder to reason about).

molsonkiko replied:
If you'd prefer something like
```py
msg = ("an array's repr should show dtype if and only if the type "
                 'of the array is NOT one of the standard types '
                 '(e.g., int32, bool, float64).')
if native_dtype in _typelessdata:
    assert 'dtype' in native_repr, msg
else:
    assert 'dtype' not in native_repr, msg
```
I see nothing wrong with that. That doesn't seem much more clear to me, but maybe I'm the only one.

seberg replied:
To me it is a bit nicer, especially if the `msg` was a comment on the branch.  But we already cleaned up the test a lot and this doesn't matter enough.

Thanks @molsonkiko it is a nice improvement/fix.

seberg reviewed (approved):
Just going to remove that "U1" logic as well, thanks for the follow up.

seberg committed:
Apply suggestions from code review
